File: abc/abc164/abc164_d.cs
// https://atcoder.jp/contests/abc164/submissions/12844007

using System;
using System.Linq;

class Program
{
    static void Main()
    {
        var S = Console.ReadLine();
        var T = new int[S.Length + 1];

        var MOD = 2019;
        var m10 = 1;
        for (int i = T.Length - 2; i >= 0; i--)
        {
            T[i] = (T[i + 1] + m10 * (S[i] - '0')) % MOD;
            m10 = m10 * 10 % MOD;
        }

        var s = new int[MOD];
        for (int i = 0; i < T.Length; i++)
        {
            s[T[i]]++;
        }
        Console.WriteLine(s.Select(x => x * (x - 1) / 2).Sum());
    }
}

File: abc/abc164/abc164_d.fs
// https://atcoder.jp/contests/abc164/submissions/12844734

let S = stdin.ReadLine().ToCharArray()
let T = Array.zeroCreate<int> (S.Length + 1)
let MOD = 2019

let rec solver m n =
    if n = -1 then ()
    else
        T.[n] <- (T.[n + 1] + m * (int S.[n] - int '0')) % MOD
        solver (m * 10 % MOD) (n - 1)

solver 1 (T.Length - 2)

T
|> Array.countBy id
|> Array.sumBy (fun x -> snd x * (snd x - 1) / 2)
|> stdout.WriteLine
